### PR TITLE
DOC: Unpin numpydoc, add nitpick exception rules

### LINKED ISSUE #10529: MNT: Undo numpydoc pin when we can
This is a follow-up of #10524 . When a new version of something (numpydoc, Sphinx, automodapi, etc) comes out upstream that fixes the `:py:class reference target not found:` warnings, we should remove this pinning:

https://github.com/astropy/astropy/blob/3a8732fd2e99653fca8e766fc72245ca98b240cd/setup.cfg#L76

If even then, some fix to our own docstrings is still needed, that fix should be applied as well.

### PR DESCRIPTION
**Pull request by pllim**
Follow up of #10524 . Fix #10529. Thanks to https://github.com/numpy/numpydoc/issues/275#issuecomment-654418024 .

### CODE REVIEW

**Comment by pllim** [comment]
So close....

File: astropy/wcs/docstrings.py [hunk 1]
```diff
@@ -824,9 +824,9 @@ def RA_DEC_ORDER(indent=0):
 """
 
 get_cdelt = """
-get_cdelt() -> ``double array[naxis]``
+get_cdelt() -> numpy.ndarray
```
Comment (pllim) [inline astropy/wcs/docstrings.py:827]: Note: Apparently linking was still attempted despite the double-backticks. This is some C-extension thingy, so I am not going to question it... 🤷

**Comment by pllim** [comment]
RTD passed without numpydoc pinning! I'll let the full CI run too, just in case.

File: docs/conf.py [hunk 2]
```diff
@@ -200,7 +200,16 @@
 # resolve.
 
 nitpicky = True
-nitpick_ignore = []
+# This list is from https://github.com/numpy/numpydoc/issues/275
+nitpick_ignore = [
```
Comment (mhvk) [inline docs/conf.py:204]: I think we should add them to nitpick_exceptions; I found it needs to be without the quotes:
```
# sphinx doing strange things with dict methods
py:class None.  Remove all items from D.
py:class a set-like object providing a view on D's items
py:class a set-like object providing a view on D's keys
py:class v, remove specified key and return the corresponding value.
py:class None.  Update D from dict/iterable E and F.
py:class an object providing a view on D's values
py:class a shallow copy of D
```

**Comment by astrofrog** [comment]
Thanks! Is there a reason the nitpick exceptions can’t live in the dedicated file?

**Comment by pllim** [comment]
> Is there a reason the nitpick exceptions can’t live in the dedicated file?

No. I wasn't even aware the is a dedicated file. Personally, I like all the settings in one place but I'll move it as requested.

File: docs/conf.py (hunk 2 above)
Comment (pllim) [reply docs/conf.py]: Done. Thanks!

**Comment by astrofrog** [comment]
I am backporting this to v4.1.x to fix some docs failures there

**Pull request merged**